keps/sig-scheduling/5004-dra-extended-resource/README.md

```diff
@@ -0,0 +1,989 @@
+        // Status of resource claims.
+        // +patchMergeKey=name
+        // +patchStrategy=merge,retainKeys
+        // +listType=map 
+        // +listMapKey=name
+        // +featureGate=DynamicResourceAllocation
+        // +optional
+        ResourceClaimStatuses []PodResourceClaimStatus `json:"resourceClaimStatuses,omitempty" patchStrategy:"merge,retainKeys" patchMergeKey:"name" protobuf:"bytes,15,rep,name=resourceClaimStatuses"`
+        // Status of extended resource claim backed by DRA.
+        // +featureGate=DynamicResourceAllocation
+        // +optional
+        ExtendedResourceClaimStatus *PodExtendedResourceClaimStatus `json:"extendedResourceClaimStatus,omitempty" protobuf:"bytes,17,opt,name=extendedResourceClaimStatus"`
+}
+```
+
+For example, if a pod has requested for foo.domain/bar, and it is
+scheduled to run on a node where foo.domain/bar was mapped to devices in a DeviceClass,
+then the pod's status is like below:
+
+```yaml
+status:
+   extendedResourceClaimStatus:
+   - names:
+     - container-name
+     - foo.domain/bar
+     - container-0-request-2
+   resourceClaimName: ccc-gpu-57999b9c4c-vpq68-gpu-8s27z
+```
+where `deviceRequest` name is "container-0-request-2", and container-name is the first container
+in the pod, foo.domain/bar is the 3rd extended resource in the container's requests.
+
+Note the validations for extendedResourceClaimStatus are different from the
+validations for resourceClaimStatuses.
+
+1. resourceClaimStatuses requires `name` must be DNS label,
+   extendedResourceClaimStatus's names' `containerName` and `RequestName` must
+   be a DNS label, while the `extendedResourceName` is not a DNS label.
+1. resourceClaimStatuses requires `name` must be one of the claim's name in the
+   pod spec. extendedResourceClaimStatus requires `containerName` must be one
+   of the container name in the pod spec, and `extendedReourceName` must be one
+   of the extended resource name in that container.
+
+### Scheduling for Extended Resource backed by DRA
+
+A new field `DynamicResources` is added to
+[`Resource`](https://github.com/kubernetes/kubernetes/blob/c81431de59a3bf516489317433a165b050322339/pkg/scheduler/framework/types.go#L798),
+it works similar to ScalarResources. It is used to keep track of the extended
+resources backed by DRA on a node, i.e. those that are advertised by `ResourceSlice`.
+
+```go
+type Resource struct {
+	MilliCPU         int64
+	Memory           int64
+	EphemeralStorage int64
+	// We store allowedPodNumber (which is Node.Status.Allocatable.Pods().Value())
+	// explicitly as int, to avoid conversions and improve performance.
+	AllowedPodNumber int
+	// ScalarResources
+	ScalarResources map[v1.ResourceName]int64
+
+	// NEW!
+	// DynamicResources: keep track of extended resources backed by DRA to device classes
+	// The map's key is the extended resource name that has at least one device
+	// class advertises it.
+	DynamicResources map[v1.ResourceName][]string
```

Comment:
"at least one" - but I think we said "exactly one" above, correct? So this can become a `map[v1.ResourceName]string`?